sdk/typescript/src/signers/signer-with-provider.ts

```diff
@@ -332,12 +357,79 @@ export abstract class SignerWithProvider implements Signer {
    * @returns total gas cost estimation
    * @throws whens fails to estimate the gas cost
    */
-  async getGasCostEstimation(...args: Parameters<SignerWithProvider['dryRunTransaction']>) {
+  async getGasCostEstimation(
+    ...args: Parameters<SignerWithProvider['dryRunTransaction']>
+  ) {
     const txEffects = await this.dryRunTransaction(...args);
     const gasEstimation = getTotalGasUsed(txEffects);
     if (typeof gasEstimation === 'undefined') {
       throw new Error('Failed to estimate the gas cost from transaction');
     }
     return gasEstimation;
   }
+
+  /**
+   * Calculates the suggested gas budget and gas estimation of a transaction, using {@link SignerWithProvider.dryRunTransaction}.
+   * In case there is an error, or dryRun is not supported by the environment it uses a default gas budget and const estimation based
+   * on the transaction kind.
+   * @param txKind The transaction kind
+   * @param txCreator A function that accepts a guessed gas budget and returns the transaction to dry run
+   * @param maxGasCoinBalance The maximum amount of the balance of an individual SUI coin owned by sender. There is no way to use multiple coins for gas at the moment so the sender has to own one with enough balance.
+   * @param budgetParams Any other parameters required to guess the gasBudget before dry running the transaction
+   * @returns An object the includes
+   * suggestedGasBudget - in case of error or insufficient gas this defaults to the 'guessed' budget but without taking into account the maxGasCoinBalance and available balance
+   * gasCostEstimation - in case of error or insufficient gas this defaults to the 'guessed' budget
+   * insufficientGas
+   * isSuccess
+   */
+  async getGasCostEstimationAndSuggestedBudget<
+    T extends GasEstimationUtils.TxKind
+  >(
+    txKind: T,
+    txCreator: (
+      gasBudgetGuess: number
+    ) => Promise<Parameters<SignerWithProvider['dryRunTransaction']>>,
+    maxGasCoinBalance: bigint,
+    ...budgetParams: GasEstimationUtils.GasBudgetGuessParams<T>
+  ) {
+    let gasCostEstimation: number;
+    let suggestedGasBudget: number;
+    let insufficientGas: boolean;
+    let isSuccess: boolean;
+    const gasBudget = GasEstimationUtils.getGasBudgetGuess(
+      txKind,
+      maxGasCoinBalance,
+      false,
+      budgetParams
+    );
+    try {
+      const dryRunResult = await this.dryRunTransaction(
+        ...(await txCreator(gasBudget))
+      );
+      isSuccess = dryRunResult.status.status === 'success';
+      gasCostEstimation =
+        (isSuccess ? getTotalGasUsed(dryRunResult) : null) ?? gasBudget;
+      suggestedGasBudget = isSuccess
+        ? dryRunResult.gasUsed.computationCost +
+          dryRunResult.gasUsed.storageCost
+        : GasEstimationUtils.getGasBudgetGuess(
+            txKind,
+            null,
+            true,
```

Comment:
I find this forError is confusing, can we derive it from if maxGasCoinBalance is null or not?

Reply:
Yeah you are right, I will change it to that. I introduced this value to notify `getGasBudgetGuess` (and make it more intentional) that we don't have dry run or something didn't work (with dry-run) so make a guess based on that (do not optimise the guess based on minimum available coins and max IndividualCoinBalance).

Probably removing it and deriving it from `maxGasCoinBalance` should be ok.